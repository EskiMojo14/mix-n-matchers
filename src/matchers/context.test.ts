/* eslint-disable @typescript-eslint/no-empty-function */
import type { Mock } from "@globals";
import { it, expect, describe, fn } from "@globals";
import type { CallInfo, Spy } from "../utils";

// Given a Jest mock function, return a minimal mock of a spy.
const createSpy = (mock: Mock): Spy => {
  const spy = function () {};

  spy.calls = {
    all() {
      const info: Array<CallInfo> = [];
      let i = 0;
      while (i < mock.mock.calls.length) {
        const returnRecord = mock.mock.results[i];
        /* eslint-disable @typescript-eslint/no-unsafe-assignment */
        info.push({
          args: mock.mock.calls[i] ?? [],
          object:
            "contexts" in mock.mock
              ? mock.mock.contexts[i]
              : (mock.mock as { instances: Array<unknown> }).instances[i],
          returnValue:
            returnRecord?.type === "return" ? returnRecord.value : undefined,
        });
        /* eslint-enable @typescript-eslint/no-unsafe-assignment */
        i++;
      }
      return info;
    },
    count() {
      return mock.mock.calls.length;
    },
  };

  return spy;
};

describe.each([
  "toHaveBeenCalledWithContext",
  "toHaveBeenLastCalledWithContext",
  "toHaveBeenNthCalledWithContext",
] as const)("%s", (calledWithContext) => {
  it("works only on spies or fn", () => {
    const mock = function mock() {};
    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(() => {
        expect(mock)[calledWithContext](3, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(() => {
        expect(mock)[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works when not called", () => {
    const mock = fn();
    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock)).not[calledWithContext](1, "foo");
      expect(mock).not[calledWithContext](1, "foo");

      expect(() => {
        expect(mock)[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock)).not[calledWithContext]("foo");
      expect(mock).not[calledWithContext]("foo");

      expect(() => {
        expect(mock)[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that doesn't match", () => {
    const mock = fn();
    mock.call("bar");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock)).not[calledWithContext](1, "foo");
      expect(mock).not[calledWithContext](1, "foo");

      expect(() => {
        expect(mock)[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock)).not[calledWithContext]("foo");
      expect(mock).not[calledWithContext]("foo");

      expect(() => {
        expect(mock)[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that doesn't match with matchers", () => {
    const mock = fn();
    mock.call("bar");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock)).not[calledWithContext](1, expect.any(Number));
      expect(mock).not[calledWithContext](1, expect.any(Number));

      expect(() => {
        expect(mock)[calledWithContext](1, expect.any(Number));
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock)).not[calledWithContext](expect.any(Number));
      expect(mock).not[calledWithContext](expect.any(Number));

      expect(() => {
        expect(mock)[calledWithContext](expect.any(Number));
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that matches", () => {
    const mock = fn();
    mock.call("foo");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock))[calledWithContext](1, "foo");
      expect(mock)[calledWithContext](1, "foo");

      expect(() => {
        expect(mock).not[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock))[calledWithContext]("foo");
      expect(mock)[calledWithContext]("foo");

      expect(() => {
        expect(mock).not[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that matches with a matcher", () => {
    const mock = fn();
    mock.call("foo");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock))[calledWithContext](1, expect.any(String));
      expect(mock)[calledWithContext](1, expect.any(String));

      expect(() => {
        expect(mock).not[calledWithContext](1, expect.any(String));
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock))[calledWithContext](expect.any(String));
      expect(mock)[calledWithContext](expect.any(String));

      expect(() => {
        expect(mock).not[calledWithContext](expect.any(String));
      }).toThrowErrorMatchingSnapshot();
    }
  });

  it("includes the custom mock name in the error message", () => {
    const mock = fn().mockName("named-mock");

    mock.call("foo");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(mock)[calledWithContext](1, "foo");

      expect(() => {
        expect(mock).not[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(mock)[calledWithContext]("foo");

      expect(() => {
        expect(mock).not[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
});
