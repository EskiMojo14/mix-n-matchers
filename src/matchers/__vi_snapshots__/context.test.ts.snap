// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toHaveBeenCalledWithContext > includes the custom mock name in the error message 1`] = `
expect(named-mock).not.toHaveBeenCalledWithContext(expected)

Expected: not "foo"
Received:     "foo"

Number of calls: 1
`;

exports[`toHaveBeenCalledWithContext > works only on spies or fn 1`] = `
expect(received).toHaveBeenCalledWithContext(expected)

Matcher error: received value must be a mock or spy function

Received has type:  function
Received has value: [Function mock2]
`;

exports[`toHaveBeenCalledWithContext > works when not called 1`] = `
expect(spy).toHaveBeenCalledWithContext(expected)

Expected: "foo"

Number of calls: 0
`;

exports[`toHaveBeenCalledWithContext > works with a context that doesn't match 1`] = `
expect(spy).toHaveBeenCalledWithContext(expected)

Expected: "foo"
Received: "bar"
Number of calls: 1
`;

exports[`toHaveBeenCalledWithContext > works with a context that doesn't match with matchers 1`] = `
expect(spy).toHaveBeenCalledWithContext(expected)

Expected: Any<Number>
Received: "bar"
Number of calls: 1
`;

exports[`toHaveBeenCalledWithContext > works with a context that matches 1`] = `
expect(spy).not.toHaveBeenCalledWithContext(expected)

Expected: not "foo"
Received:     "foo"

Number of calls: 1
`;

exports[`toHaveBeenCalledWithContext > works with a context that matches with a matcher 1`] = `
expect(spy).not.toHaveBeenCalledWithContext(expected)

Expected: not Any<String>
Received:     "foo"

Number of calls: 1
`;

exports[`toHaveBeenLastCalledWithContext > includes the custom mock name in the error message 1`] = `
expect(named-mock).not.toHaveBeenLastCalledWithContext(expected)

Expected: not "foo"

Number of calls: 1
`;

exports[`toHaveBeenLastCalledWithContext > works only on spies or fn 1`] = `
expect(received).toHaveBeenLastCalledWithContext(expected)

Matcher error: received value must be a mock or spy function

Received has type:  function
Received has value: [Function mock2]
`;

exports[`toHaveBeenLastCalledWithContext > works when not called 1`] = `
expect(spy).toHaveBeenLastCalledWithContext(expected)

Expected: "foo"

Number of calls: 0
`;

exports[`toHaveBeenLastCalledWithContext > works with a context that doesn't match 1`] = `
expect(spy).toHaveBeenLastCalledWithContext(expected)

Expected: "foo"
Received: "bar"
Number of calls: 1
`;

exports[`toHaveBeenLastCalledWithContext > works with a context that doesn't match with matchers 1`] = `
expect(spy).toHaveBeenLastCalledWithContext(expected)

Expected: Any<Number>
Received: "bar"
Number of calls: 1
`;

exports[`toHaveBeenLastCalledWithContext > works with a context that matches 1`] = `
expect(spy).not.toHaveBeenLastCalledWithContext(expected)

Expected: not "foo"

Number of calls: 1
`;

exports[`toHaveBeenLastCalledWithContext > works with a context that matches with a matcher 1`] = `
expect(spy).not.toHaveBeenLastCalledWithContext(expected)

Expected: not Any<String>
Received:     "foo"

Number of calls: 1
`;

exports[`toHaveBeenNthCalledWithContext > includes the custom mock name in the error message 1`] = `
expect(named-mock).not.toHaveBeenNthCalledWithContext(n, expected)

n: 1
Expected: not "foo"

Number of calls: 1
`;

exports[`toHaveBeenNthCalledWithContext > works only on spies or fn 1`] = `
expect(received).toHaveBeenNthCalledWithContext(n, expected)

Matcher error: received value must be a mock or spy function

Received has type:  function
Received has value: [Function mock2]
`;

exports[`toHaveBeenNthCalledWithContext > works when not called 1`] = `
expect(spy).toHaveBeenNthCalledWithContext(n, expected)

n: 1
Expected: "foo"

Number of calls: 0
`;

exports[`toHaveBeenNthCalledWithContext > works with a context that doesn't match 1`] = `
expect(spy).toHaveBeenNthCalledWithContext(n, expected)

n: 1
Expected: "foo"
Received: "bar"
Number of calls: 1
`;

exports[`toHaveBeenNthCalledWithContext > works with a context that doesn't match with matchers 1`] = `
expect(spy).toHaveBeenNthCalledWithContext(n, expected)

n: 1
Expected: Any<Number>
Received: "bar"
Number of calls: 1
`;

exports[`toHaveBeenNthCalledWithContext > works with a context that matches 1`] = `
expect(spy).not.toHaveBeenNthCalledWithContext(n, expected)

n: 1
Expected: not "foo"

Number of calls: 1
`;

exports[`toHaveBeenNthCalledWithContext > works with a context that matches with a matcher 1`] = `
expect(spy).not.toHaveBeenNthCalledWithContext(n, expected)

n: 1
Expected: not Any<String>
Received:     "foo"

Number of calls: 1
`;
